bundles/framework/statehandler/session-methods.js

```diff
@@ -47,15 +50,37 @@ Oskari.clazz.category(
                 me.setSessionExpiring(minutes);
                 popup.close(true);
             });
+            logoutButton.setTitle(logoutButtonMessage);
+            logoutButton.setHandler(function () {
+                jQuery('#loginbar').find("form[action='/logout']").submit();
+            });
 
             return {
                 show: function () {
-                    popup.show(notifyTitle, notifyMessage, [extendButton]);
-
-                    expireTimeout = setTimeout(function () {
-                        popup.show(expiredTitle, expiredMessage);
-                        popup.makeModal();
-                    }, 60 * 1000);
+                    const expireIn = 60; // Expire time in seconds
+                    popup.show(notifyTitle, notifyMessage + '<br />' + expiresInMessage.replace('{expires}', (expireIn)), [logoutButton, extendButton]);
+                    // Using Date for more accurate countdown (instead of just using setTimeout or setInterval)
+                    const start = Date.now();
+                    let diff;
+                    let seconds;
+                    function timer () {
+                        diff = expireIn - (((Date.now() - start) / 1000) | 0);
+                        seconds = (diff % 60) | 0;
+                        seconds = seconds < 10 ? 0 + seconds : seconds;
+                        if (seconds === 0) {
+                            clearInterval(interval);
+                            popup.show(expiredTitle, expiredMessage);
+                            popup.makeModal();
+                            if (Oskari.user().isLoggedIn()) {
+                                jQuery('#loginbar').find("form[action='/logout']").submit();
+                            } else {
+                                location.reload();
+                            }
+                        } else {
+                            popup.setContent(notifyMessage + '<br />' + expiresInMessage.replace('{expires}', seconds));
+                        }
+                    }
+                    let interval = setInterval(timer, 1000);
```

Comment:
It would be more readable if variable `interval` was defined before it's used on line 71.